internal/app/service/service.go

```diff
@@ -0,0 +1,60 @@
+package service
+
+import (
+	"errors"
+
+	e "github.com/patraden/ya-practicum-go-shortly/internal/app/errors"
+	"github.com/patraden/ya-practicum-go-shortly/internal/app/repository"
+	"github.com/patraden/ya-practicum-go-shortly/internal/app/urlgenerator"
+)
+
+type ShortenerService struct {
+	URLShortener
+	repo         repository.URLRepository
+	urlGenerator urlgenerator.URLGenerator
+	retries      int
+}
+
+func NewShortenerService() *ShortenerService {
+	return &ShortenerService{
+		repo:         repository.NewMapURLRepository(),
+		urlGenerator: urlgenerator.NewRandURLGenerator(8),
+		retries:      1000,
+	}
+}
+
+func (s *ShortenerService) ShortenURL(longURL string) (string, error) {
+	shortURL := s.urlGenerator.GenerateURL(longURL)
+
+	tries := 0
+	_, err := s.repo.AddURL(shortURL, longURL)
+	for errors.Is(err, e.ErrExists) && tries < s.retries {
+		shortURL = s.urlGenerator.GenerateURL(longURL)
+		_, err = s.repo.AddURL(shortURL, longURL)
+		tries += 1
```

Comment:
[SUGGESTION] If I were, I would use `backoff`.
https://pkg.go.dev/github.com/cenkalti/backoff/v4#section-readme

Reply:
[CHANGE REQUEST] It doesn't make sense for `s.repo.AddURL`.